### PR TITLE
ARROW-9588: [C++] Partially support building with clang in an MSVC setting

### PR DESCRIPTION
**Pull request by xhochy**
This is only half the work, thus we don't have a test for this yet but I wanted to upstream this to have a local diff less lying around.

### CODE REVIEW

**Comment by kou** [comment]
It seems that `@github-actions autotune` doesn't support fixing CMake format yet: https://github.com/apache/arrow/blob/master/.github/workflows/comment_bot.yml#L87

File: cpp/src/arrow/python/type_traits.h [hunk 1]
```diff
@@ -105,7 +105,7 @@ struct npy_traits<NPY_FLOAT32> {
   using TypeClass = FloatType;
   using BuilderClass = FloatBuilder;
 
-  static constexpr float na_sentinel = NAN;
+  static constexpr float na_sentinel = std::numeric_limits<float>::quiet_NaN();
```
Comment (pitrou) [inline cpp/src/arrow/python/type_traits.h:108]: Can you explain why this change is necessary? The `NAN` macro is standard: https://en.cppreference.com/w/cpp/numeric/math/NAN

Comment (xhochy) [reply cpp/src/arrow/python/type_traits.h:110]: Added a comment, the issue is that the Windows implementation of the macro is not fully compatible with clang's stictness.

Comment (pitrou) [reply cpp/src/arrow/python/type_traits.h:110]: Thank you. Just: "division" (typo)

**Review by pitrou — APPROVED** [review]
+1
